app/models.py

```diff
@@ -0,0 +1,607 @@
+                          BadSignature, SignatureExpired)
+from validators import url
+from werkzeug.exceptions import NotFound
+from werkzeug.security import generate_password_hash, check_password_hash
+
+from app import db
+from app.api.custom_exceptions import (ValidationException,
+                                       UrlValidationException)
+from app.api.errors import bad_request
+from app.api.shortener import Shortener
+
+
+class AnonymousUser(AnonymousUserMixin):
+    @staticmethod
+    def get_anonymous_user():
+        """
+        fetches and returns an anonymous user instance
+         from the database
+        :return AnonymousUser object: 
+        """
+        anonymous_user = User.get_by_username("AnonymousUser")
+        if anonymous_user:
+            return anonymous_user
+        return AnonymousUser.create_anonymous_user()
+
+    @staticmethod
+    def create_anonymous_user():
+        """
+        creates an AnonymousUser Object and save it into the database
+        :return AnonymousUser Object: 
+        """
+        anonymous_user = User(username='AnonymousUser', lastname="anonymous",
+                              firstname='anonymous', email='anonymous')
+        db.session.add(anonymous_user)
+        db.session.commit()
+        return anonymous_user
+
+    @staticmethod
+    def set_anonymous():
+        """
+        sets the global object current_user attribute to
+        an AnonymousUser instance
+        """
+        if g.current_user.is_anonymous:
+            g.current_user = AnonymousUser.get_anonymous_user()
+
+"""
+ this defines the table that holds the relationship between a User Model 
+ and the Url model
+"""
+user_url = db.Table('user_url',
+                    db.Column('user_id', db.Integer,
+                              db.ForeignKey('users.id')),
+                    db.Column('urls_id', db.Integer,
+                              db.ForeignKey('url.id')))
+
+
+class User(db.Model, UserMixin):
+    """
+    This class defines the user table schema and also handles user object related
+    functionalities.
+    """
+    __tablename__ = 'users'
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(20), unique=True, index=True,
+                         nullable=False)
+    lastname = db.Column(db.String(20), nullable=False)
+    firstname = db.Column(db.String(20), nullable=False)
+    email = db.Column(db.String(30), unique=True, index=True, nullable=False)
+    password_hash = db.Column(db.String(128))
+    is_admin = db.Column(db.Boolean, default=False)
+    date_added = db.Column(db.DateTime, default=datetime.utcnow())
+    url = db.relationship("Url", secondary=user_url, backref='users',
+                          lazy='dynamic')
+    short_url = db.relationship("ShortenUrl",  backref='users',
+                                lazy='dynamic')
+    token = db.relationship("Token", backref="users", lazy="dynamic")
```

Comment:
A token can't be related to several users, its a one-to-one relationship from the perspective of the token. the backref gives me the ability to fetch information from both sides of the table. From the users model, i can fetch all the tokens that has been assigned to a user, and from the token model, i can fetch the user using that particular token. 
`NB`: based on the insight stated above about the violation of the stateless attribute of an api, the token table has been modified to hold blacklisted tokens in order to avoid their future usage to access endpoints.